api/oc_push.c

```diff
@@ -0,0 +1,2605 @@
+
+					/* receiver:rts[] */
+					oc_rep_open_array(receivers, rts);
+					for (int j=0; j < (int)oc_string_array_get_allocated_size(recv_obj->rts); j++)
+					{
+						oc_rep_add_text_string(rts, oc_string_array_get_item(recv_obj->rts, j));
+					}
+					oc_rep_close_array(receivers, rts);
+
+					/* == close object == */
+					oc_rep_object_array_end_item(receivers);
+
+					recv_obj = recv_obj->next;
+				}
+
+				break;
+			}
+
+			recvs_instance = recvs_instance->next;
+		}
+		oc_rep_close_array(root, receivers);
+		break;
+	default:
+		break;
+	}
+	oc_rep_end_root_object();
+
+	oc_send_response(request, result);
+
+	return;
+}
+
+
+
+/*
+ * @brief				get length of payload list (oc_rep_t list)
+ *
+ * @param obj_list	payload list
+ * @return				number of payload list member
+ */
+static int _get_obj_array_len(oc_rep_t *obj_list)
+{
+	int n = 0;
+	oc_rep_t *obj = obj_list;
+
+	while (obj)
+	{
+		n++;
+		obj = obj->next;
+	}
+
+	return n;
+}
+
+
+
+/*
+ * @brief					purge app resource and resource representation container
+ * 							accessed through `uri` in device whose index is `device_index`
+ *
+ * @param uri				URI to app resource to be purged
+ * @param device_index	index of device where the target resource resides
+ */
+static void _purge_pushd_rsc(oc_string_t *uri, size_t device_index)
+{
+	oc_resource_t *pushd_rsc = oc_ri_get_app_resource_by_uri(oc_string(*uri), oc_string_len(*uri), device_index);
+	oc_pushd_rsc_rep_t *pushd_rsc_rep = _find_pushd_rsc_rep_by_uri(uri, device_index);
+
+	if (pushd_rsc_rep)
+	{
+		/* step 1. purge `rep` */
+		oc_rep_set_pool(&rep_instance_memb);
+		oc_free_rep(pushd_rsc_rep->rep);
+
+		/* step 2. remove pushed resource representation from `pushed_rsc_rep_list` */
+		oc_list_remove(pushd_rsc_rep_list, pushd_rsc_rep);
+		oc_memb_free(&pushd_rsc_rep_instance_memb, pushd_rsc_rep);
+	}
+	else
+	{
+		OC_PUSH_ERR("can't find resource representation for pushed resource (%s)...", oc_string(*uri));
+		return;
+	}
+
+	if (pushd_rsc)
+	{
+		/* step 3. remove pushed Resource from `app_resources` */
+		OC_PUSH_DBG("purge pushed resource (%s)...", oc_string(*uri));
+		oc_delete_resource(pushd_rsc);
+	}
+	else
+	{
+		OC_PUSH_ERR("can't find pushed resource (%s)...", oc_string(*uri));
+		return;
+	}
+
+	return;
+}
+
+
+
+/*
+ * @brief				create app resource and `oc_pushd_rsc_t` corresponding to pushed resource
+ *
+ * @param recv_obj	receiver object that points pushed resource
+ * @param resource	Push Receiver resource
+ */
+static bool _create_pushd_rsc(oc_recv_t *recv_obj, oc_resource_t *resource)
+{
+	bool result = true;
+
+	/* create Push Receiver Resource */
+	oc_resource_t *pushd_rsc = oc_new_resource("Pushed Resource", oc_string(recv_obj->receiveruri), 1, resource->device);
+
+	if (pushd_rsc)
+	{
+		/*
+		 * XXX, if a resource binds empty resource type (""), when a client retrieve this it may receive weird value...
+		 */
+		oc_resource_bind_resource_type(pushd_rsc, " ");
```

Comment:
Why the empty space " ", isn't there some designated resource type in the specification?

@jkralik